Add clipboard button

KTibow opened the pull request:
# **Moved to #8411**



<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change

<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This adds a button next to the entity ID that copies the ID to the clipboard.

## Type of change

<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->

```yaml

```

## Additional information

<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
![image](https://user-images.githubusercontent.com/10727862/98732483-131fac80-2354-11eb-9fcc-73e07c490872.png)

- This PR fixes or closes issue: fixes #
- This PR is related to issue: #7599
- Link to documentation pull request:

## Checklist

<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io

KTibow committed:
Add clipboard button

spacegaier commented:
Can you please post a screenshot.

KTibow commented:
Okay... Let me get my dev instance running.

KTibow commented:
![image](https://user-images.githubusercontent.com/10727862/98444348-0dad3280-20c6-11eb-9fd8-1c68e04da3b8.png)
I can adjust the padding or whatever if you want. Originally the button was on the right but now it's next to the more info button.

KTibow commented:
ooops made a git oopsi

spacegaier commented:
> 
> 
> ![image](https://user-images.githubusercontent.com/10727862/98444348-0dad3280-20c6-11eb-9fd8-1c68e04da3b8.png)
> I can adjust the padding or whatever if you want. Originally the button was on the right but now it's next to the more info button.

Yeah, I think currently the new button is not justified between the more-info button and the hyperlink. Needs to move slightly to the left I think.

KTibow commented:
@spacegaier What would be a good way to do that? I know using `&nbsp;` is a bad practice. Would adding `12px` for `margin-right` work?

spacegaier commented:
> 
> 
> @spacegaier What would be a good way to do that? I know using `&nbsp;` is a bad practice. Would adding `12px` for `margin-right` work?

Not sure what the best way is (not a CSS guru myself). Perhaps putting all three elements into a flex box might work and sort that out. The more manual approach would be so set a CSS class for the second icon that removes the left margin if that is causing the too wide gap to the first icon.

bramkragten commented on src/panels/developer-tools/state/developer-tools-state.js:
```diff
@@ -254,6 +263,11 @@ class HaPanelDevState extends EventsMixin(LocalizeMixin(PolymerElement)) {
 
+  copyEntity(ev) {
+    ev.preventDefault();
+    navigator.clipboard.writeText(ev.model.entity.entity_id);
```
There is not full browser support for this API, use <https://github.com/home-assistant/frontend/blob/dev/src/common/util/copy-clipboard.ts> instead. We could add `navigator.clipboard.writeText` in that helper if we check for support first.

KTibow replied:
Fixed and improved.

bramkragten commented:
Make sure there is enough space between the buttons so you can touch it on mobile. Check the guidelines: https://material.io/develop/web/components/buttons/icon-buttons

Related: <https://github.com/home-assistant/frontend/pull/4259>

KTibow committed:
Use copyToClipboard instead of navigator.clipboard

bramkragten commented:
You should rebase, not merge everything in your branch.

Misiu commented on src/panels/developer-tools/state/developer-tools-state.js:
```diff
@@ -169,6 +173,12 @@ class HaPanelDevState extends EventsMixin(LocalizeMixin(PolymerElement)) {
+                on-click="copyEntity"
+                alt="[[localize('ui.panel.developer-tools.tabs.states.copy_id')]]"
+                title="[[localize('ui.panel.developer-tools.tabs.states.copy_id')]]"
+                path="[[clipboardOutlineIcon()]]"
```
I think the icon can be put as a property. Like here: https://github.com/home-assistant/frontend/blob/c22022856616c10efa986dfc6cfe97854c1195e1/src/dialogs/more-info/controls/more-info-weather.ts#L86

```suggestion
                .path=${mdiClipboardTextMultipleOutline}
```

and `clipboardOutlineIcon()` can be removed

Misiu replied:
Sorry, I just noticed this is still using Polymer. NVM.

KTibow replied:
Getting on the build. Do you think that I should do that for the other icon, or keep the PR small?

Misiu replied:
@KTibow sorry for the trouble, leave this as it. I noticed that this is in Polymer, so my initial idea won't work 😕

KTibow replied:
Okay.

KTibow commented:
Just curious: Why did #4259 cause lag? How can I prevent this PR from causing lag?

bramkragten commented:
Because it had a dropdown menu and button on every row, that needed to be rendered/calculated, since this is just 1 icon on every row it should be ok.

bramkragten commented on src/common/util/copy-clipboard.ts:
```diff
@@ -1,8 +1,12 @@
-  document.execCommand("copy");
-  document.body.removeChild(el);
+  if (navigator.clipboard) {
+    navigator.clipboard.writeText(str);
```
This will require permission, that is also not supported in all browsers (Firefox, Safari) we should check if this is successful and otherwise fallback

KTibow replied:
It shouldn't require permission as long as it's user-initiated. It works just fine for me in the latest version of Firefox, don't have a Mac so can't test safari. Do you mean that the `if (navigator.clipboard) {` might not catch it? I put a fallback in anyway.

KTibow committed:
Split out clipboard fallback

The clipboard fallback triggers if navigator.clipboard is falsey, or if
an error occurs using navigator.clipboard.
Thanks for helping improve the code Bram!

Co-Authored-By: Bram Kragten <bramkragten@users.noreply.github.com>

bramkragten commented on src/common/util/copy-clipboard.ts:
```diff
@@ -1,8 +1,16 @@
+
+export const copyToClipboard = (str) => {
+  if (navigator.clipboard) {
+    navigator.clipboard.writeText(str).then(() => {}, fallbackCopy);
```
```suggestion
    navigator.clipboard.writeText(str).catch(() => fallbackCopy(str));
```

bramkragten commented on src/common/util/copy-clipboard.ts:
```diff
@@ -1,8 +1,16 @@
-export const copyToClipboard = (str) => {
+function fallbackCopy(str) {
```
```suggestion
const fallbackCopy = (str) => {
```

bramkragten commented:
We can also just use this: https://github.com/feross/clipboard-copy/blob/master/index.js

KTibow committed:
Just use the dev version of copy to clipboard
